src/AutoMapper/Mappers/Internal/CollectionMapperExpressionFactory.cs

```diff
@@ -46,7 +47,7 @@ public static Expression MapCollectionExpression(IConfigurationProvider configur
                     Assign(passedDestination, destExpression),
                     assignNewExpression,
                     Call(destination, clearMethod),
-                    ToType(mapExpr, createInstance.Type)
```

Comment:
Stupid question.  Wouldn't this be sufficient not to cast the result back?  Not sure why anything else would be needed.  The problem wasn't that collection was a List when passed down it's cause it was casted as such instead of ICollection.